larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,299 @@
+        "VariableID", variableID));
+
+    double threshold(0.);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "Threshold", threshold));
+
+    int leftDaughterID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "LeftDaughterID", leftDaughterID));
+
+    int rightDaughterID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "RightDaughterID", rightDaughterID));
+
+    int parentID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "ParentID", parentID));
+
+    bool outcome(false);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "Outcome", outcome));
+
+    bool isLeaf(leftDaughterID == -1 ? true : false);
+
+    const Node *pNode = new Node(nodeID, parentID, leftDaughterID, rightDaughterID, isLeaf, threshold, variableID, outcome);
+    idToNodeMap.insert(IDToNodeMap::value_type(nodeID, pNode));
+
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+AdaBoostDecisionTree::Node::Node(const int &nodeID, const int &parentNodeID, const int &leftChildNodeID, const int &rightChildNodeID, const bool &isLeaf, const double &threshold, const int &variableID, const bool outcome) : 
+    m_nodeID(nodeID),
+    m_parentNodeID(parentNodeID),
+    m_leftChildNodeID(leftChildNodeID),
+    m_rightChildNodeID(rightChildNodeID),
+    m_isLeaf(isLeaf),
+    m_threshold(threshold),
+    m_variableID(variableID),
+    m_outcome(outcome)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+AdaBoostDecisionTree::WeakClassifier::WeakClassifier(IDToNodeMap &idToNodeMap, const double &weight, const int &treeID) : 
+    m_idToNodeMap(idToNodeMap),
+    m_weight(weight),
+    m_treeID(treeID)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AdaBoostDecisionTree::WeakClassifier::Predict(const DoubleVector &features) const 
+{   
+    return this->EvaluateNode(0, features); 
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AdaBoostDecisionTree::WeakClassifier::EvaluateNode(const int &nodeID, const DoubleVector &features) const 
```

Comment:
No need to have const int by reference, just const by value for built-ins.

Reply:
Done.  Am I correct in assuming strings are not built-in types?

Reply:
Definitely worth passing string arguments by const reference (character type is the basic type).